### PR TITLE
feat: serializeJson cheatcode

### LINKED ISSUE #5745: serializeJson cheatcode
### Component

Forge

### Describe the feature you would like

Add a new `serializeJson(string calldata objectKey, string calldata value)` cheatcode (naming might not be the best), which adds the provided json string to [`serialized_jsons`](https://github.com/foundry-rs/foundry/blob/master/crates/evm/src/executor/inspector/cheatcodes/mod.rs#L171) directly as a new object.

Currently it is possible to "create" a new serialized json in the state by specifying an `id` and a `key`, but it is not possible to start from an existing json string. `serializeJson` should just insert the value to the `serialized_jsons` or overwrite with value if the id already exists.

If this makes sense and there is interest I can try to implement it :)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vdrg**
Adds a new `serializeJson` cheatcode which receives an id and a json object string, and internally adds the object to the `serialized_jsons` map or overwrites an id if it already exists.

Closes #5745 

## Motivation

Currently there is no way to start from an existing json string and manipulate it with the `serializeX` cheatcodes (unless you do something hacky like using a nested object and introducing some abstractions to operate over it...).

To be honest, I think it would be better to have "pure" setter cheatcodes that operate over a json string and return the modified version without changing the internal state. But it would need a whole bunch of new functions and could become confusing. Anyways, this is a different discussion (let me know if something like this would actually make sense, I can try to implement it too). **Update**: I created this issue to discuss: #5764 

## Solution

The `value_key` argument of the `serialize_json` function is now an `Option<&str>`. When `value_key` is Some, the behavior stays the same as before. When it is None, the provided string is parsed into a BTree<String, Value> and inserted for the provided `object_key`.

Also, this is my first ever Rust PR 😱 so I wasn't sure about the best way of doing those nested if lets... maybe `match` would be better here? All feedback is welcome. 

**Update**: I just saw that this has conflicts with the breaking changes in #4602 . Any ideas on a different name for this function? Maybe `serializeJsonFromString`?

### CODE REVIEW

**Review by Evalir — APPROVED** [review]
actually looks pretty well implemented to me— @mds1 wdyt?

File: testdata/cheats/Json.t.sol [hunk 1]
```diff
@@ -240,6 +240,17 @@ contract WriteJsonTest is DSTest {
         vm.removeFile(path);
     }
 
+    function test_serializeRootObject() public {
```
Comment (Evalir) [inline testdata/cheats/Json.t.sol:243]: let's maybe add a comment to this test linking to the relevant issue and how this functionality was added in this PR

Comment (vdrg) [reply testdata/cheats/Json.t.sol:245]: done!

**Comment by mds1** [comment]
The PR lgtm, thanks!

> **Update**: I just saw that this has conflicts with the breaking changes in https://github.com/foundry-rs/foundry/pull/4602 . Any ideas on a different name for this function? Maybe `serializeJsonFromString`?

But @Evalir let's make sure to get @mattsse / @odyslam to check this out before merging to make sure there's no issues/conflicts between here and #4602—that one has been open a while but does seem like we should get it merged. 

Maybe the cheat in this method should be named something like `serializeWith` or `serializeFrom`

**Comment by vdrg** [comment]
> Maybe the cheat in this method should be named something like `serializeWith` or `serializeFrom`

I've been thinking about this... maybe having both `serializeJson(objectKey, json)` and `serializeJson(objectKey, key, json)` is fine and actually makes sense? Because in both cases we are doing the same thing, it's just that in this one the key being "null" means that we are serializing the root object.

**Comment by mds1** [comment]
Ah #4602 isn't merged yet since it's a breaking change: https://github.com/foundry-rs/foundry/pull/4602#issuecomment-1640908103

---

> I've been thinking about this... maybe having both `serializeJson(objectKey, json)` and `serializeJson(objectKey, key, json)` is fine and actually makes sense? Because in both cases we are doing the same thing, it's just that in this one the key being "null" means that we are serializing the root object.

Hm that seems reasonable to me, @Evalir @mattsse wdyt?

**Comment by odyslam** [comment]
Looks good to me and a much needed feature

**Comment by mattsse** [comment]
> Ah #4602 isn't merged yet since it's a breaking change: [#4602 (comment)](https://github.com/foundry-rs/foundry/pull/4602#issuecomment-1640908103)
> 
> > I've been thinking about this... maybe having both `serializeJson(objectKey, json)` and `serializeJson(objectKey, key, json)` is fine and actually makes sense? Because in both cases we are doing the same thing, it's just that in this one the key being "null" means that we are serializing the root object.
> 
> Hm that seems reasonable to me, @Evalir @mattsse wdyt?

yeah sounds reasonable to me

so no blockers?, supportive of this change

**Comment by mds1** [comment]
yep seems there's no blockers to me! once merged we'll just need book + forge-std Vm.sol updates too if you don't mind @vdrg :)

**Comment by vdrg** [comment]
Of course 🫡

**Review by Evalir — APPROVED** [review]
sweet! just catching up—happy to merge this as-is, the suggested reasons make sense to me!

**Comment by Evalir** [comment]
@vdrg merged! let's make sure we update the book and forge std :)

**Comment by vdrg** [comment]
Sure! Will prob do over the weekend

**Pull request merged**